### PR TITLE
feat(skore): Add `data_source="both"` to RocCurveDisplay for EstimatorReports

### LINKED ISSUE #2147: enh(skore): Implement option data_source="both" in RocCurveDisplay for the EstimatorReport
### What would you like to say?

Part of #1874.  

The two curves should be in the same plot.  

For the frame, we have two options.
As a reminder, today:

```python
from sklearn.datasets import load_breast_cancer
from sklearn.linear_model import LogisticRegression
from skore import train_test_split
from skore import EstimatorReport
X, y = load_breast_cancer(return_X_y=True)
split_data = train_test_split(X=X, y=y, random_state=0, as_dict=True)
classifier = LogisticRegression(max_iter=10_000)
report = EstimatorReport(classifier, **split_data)
display = report.metrics.roc()
display.frame()
```

outputs:  

Image

**option 1**  

the display.frame() would actually output two dataframes instead of 1. Here, in this display, what's the most interesting would be the plot, and not the frame.
df_train, df_test = report.metrics.roc_auc(data_set = "both").frame()

**option 2**  

We would operate some kind of merge, where the thresholds from both train and test would be used. Cf the following concatenation:  

Image

I personnally perfer option 2, to keep the output of `report.metrics.precision_recall().frame()` similar whatever the options selected in `precision_recall`.  

The decision taken here should be consistant with the one taken in #2148, therefore let's keep the conversation on the other issue.

### PR DESCRIPTION
**Pull request by auguste-probabl**
Lays the groundwork for adding this parameter to the other displays.

Closes #2147

Binary classification:

```python
from sklearn.datasets import load_breast_cancer
from sklearn.linear_model import LogisticRegression
from skore import train_test_split
from skore import EstimatorReport
X, y = load_breast_cancer(return_X_y=True)
split_data = train_test_split(X=X, y=y, random_state=0, as_dict=True)
classifier = LogisticRegression(max_iter=10_000)
report = EstimatorReport(classifier, **split_data)
display = report.metrics.roc(data_source="both")
display.plot()
```

image

Multiclass classification:

```python
from sklearn.datasets import make_classification
from sklearn.linear_model import LogisticRegression
from skore import train_test_split
from skore import EstimatorReport
X, y = make_classification(n_classes=3, n_informative=10)
split_data = train_test_split(X=X, y=y, random_state=0, as_dict=True)
classifier = LogisticRegression(max_iter=10_000)
report = EstimatorReport(classifier, **split_data)
display = report.metrics.roc(data_source="both")
display.plot()
```

image

### CODE REVIEW

File: skore/src/skore/_sklearn/_plot/metrics/roc_curve.py [hunk 1]
```diff
@@ -193,24 +193,42 @@ def _plot_single_estimator(
...
-                line_kwargs["label"] = (
-                    f"{self.data_source.title()} set "
-                    f"(AUC = {self.roc_auc['roc_auc'].item():0.2f})"
-                )
-            else:  # data_source in (None, "X_y")
-                line_kwargs["label"] = f"AUC = {self.roc_auc['roc_auc'].item():0.2f}"
-
             line_kwargs_validated = _validate_style_kwargs(
                 line_kwargs, roc_curve_kwargs[0]
             )
```
Comment (auguste-probabl) [inline skore/src/skore/_sklearn/_plot/metrics/roc_curve.py:205]: `add_line_binary` and `add_line_multiclass` are both defined inside `_plot_single_estimator`. I don't love it but didn't find a better way.

File: skore/src/skore/_sklearn/_plot/metrics/precision_recall_curve.py [hunk 2]
```diff
@@ -117,7 +117,7 @@ def __init__(
         precision_recall: DataFrame,
         average_precision: DataFrame,
         pos_label: PositiveLabel | None,
-        data_source: Literal["train", "test", "X_y"],
+        data_source: Literal["train", "test", "X_y"] | Literal["both"],
```
Comment (thomass-dev) [inline skore/src/skore/_sklearn/_plot/metrics/precision_recall_curve.py]:
```suggestion
        data_source: Literal["train", "test", "X_y", "both"],
```

File: skore/tests/unit/displays/roc_curve/test_estimator.py [hunk 3]
```diff
@@ -174,6 +177,26 @@ def test_data_source_multiclass_classification(
...
+    display.plot()
+    for class_label in estimator.classes_:
+        train_auc = display.roc_auc.query(
+            f"label == {class_label} & data_source == 'train'"
+        )["roc_auc"].item()
+        test_auc = display.roc_auc.query(
+            f"label == {class_label} & data_source == 'test'"
+        )["roc_auc"].item()
+
+        line_labels = {line.get_label() for line in display.lines_}
```
Comment (thomass-dev) [inline skore/tests/unit/displays/roc_curve/test_estimator.py]: You shouldn't recompute it for each label. Please move it between lines 181 and 182.

File: skore/src/skore/_sklearn/_estimator/metrics_accessor.py [hunk 4]
```diff
@@ -1691,17 +1691,103 @@ def _get_display(
...
+            *,
+            X,
+            y_true,
+            data_source,
+            data_source_hash,
+            cache=self._parent._cache,
+            estimator_hash=int(self._parent._hash),
+            estimator=self._parent.estimator_,
+            response_method=response_method,
+            pos_label=pos_label,
```
Comment (thomass-dev) [inline skore/src/skore/_sklearn/_estimator/metrics_accessor.py:1707]: For better readability, you should move the function outside.

File: skore/src/skore/_sklearn/_plot/metrics/roc_curve.py [hunk 5]
```diff
@@ -193,24 +199,42 @@ def _plot_single_estimator(
...
 
-            (line,) = self.ax_.plot(
-                self.roc_curve["fpr"],
-                self.roc_curve["tpr"],
-                **line_kwargs_validated,
-            )
-            lines.append(line)
+            def add_line_binary(
+                ax, lines, *, data_source, line_kwargs=line_kwargs_validated
```
Comment (GaetandeCast) [inline skore/src/skore/_sklearn/_plot/metrics/roc_curve.py]: No need to have `ax` as a parameter since it can be accessed with `self.ax_` directly. Same in `add_line_multiclass`.

File: skore/src/skore/_sklearn/_plot/metrics/roc_curve.py (hunk 1 above)
Comment (GaetandeCast) [reply skore/src/skore/_sklearn/_plot/metrics/roc_curve.py:205]: One way would be to fuse the two functions in a _add_line() function defined outside the class. But that would require a good bit of (re)factoring.

File: skore/src/skore/_sklearn/_plot/metrics/roc_curve.py (hunk 5 above)
Comment (auguste-probabl) [inline skore/src/skore/_sklearn/_plot/metrics/roc_curve.py]: add types

File: skore/src/skore/_sklearn/_estimator/metrics_accessor.py (hunk 4 above)
Comment (auguste-probabl) [reply skore/src/skore/_sklearn/_estimator/metrics_accessor.py:1707]: since it's a helper function that is only used here (for now) I would rather leave it there
